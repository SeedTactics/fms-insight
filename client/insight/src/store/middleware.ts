/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
export enum PledgeStatus {
  Starting = "Pledge_Starting",
  Completed = "Pledge_Completed",
  Error = "Pledge_Error",
}

export type Pledge<T> =
  | { status: PledgeStatus.Starting }
  | { status: PledgeStatus.Completed; result: T }
  | { status: PledgeStatus.Error; error: Error };

export type PledgeToPromise<AP> = {
  [P in keyof AP]: "pledge" extends P ? (AP[P] extends Pledge<infer R> ? Promise<R> : AP[P]) : AP[P];
};

interface WithPromise<R> {
  pledge: Promise<R>;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function hasPledge<R>(obj: any): obj is WithPromise<R> {
  return "pledge" in obj && obj.pledge && obj.pledge instanceof Promise;
}

function pledgeMiddleware<A>(dispatch: (a: A) => void): (action: PledgeToPromise<A>) => void {
  return (action) => {
    if (hasPledge(action)) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const anyAction: any = action;
      dispatch({ ...anyAction, pledge: { status: PledgeStatus.Starting } });
      action.pledge
        .then((r) => {
          dispatch({
            ...anyAction,
            pledge: { status: PledgeStatus.Completed, result: r },
          });
        })
        .catch((e: Error) => {
          dispatch({
            ...anyAction,
            pledge: { status: PledgeStatus.Error, error: e },
          });
        });
    } else {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return dispatch(action as any);
    }
  };
}

function arrayMiddleware<A>(dispatch: (a: A) => void): (action: A | ReadonlyArray<A>) => void {
  return (action) => {
    if (action instanceof Array) {
      action.map(dispatch);
    } else {
      return dispatch(action as A);
    }
  };
}

export type ActionBeforeMiddleware<A> = PledgeToPromise<A> | ReadonlyArray<PledgeToPromise<A>>;

export function middleware<A>(dispatch: (a: A) => void): (action: ActionBeforeMiddleware<A>) => void {
  return arrayMiddleware(pledgeMiddleware(dispatch));
}
