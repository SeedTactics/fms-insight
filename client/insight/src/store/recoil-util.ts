/* Copyright (c) 2020, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from "react";
import { RecoilState, useRecoilState, useSetRecoilState } from "recoil";
import produce, { Draft, Immutable } from "immer";

export function useRecoilStateDraft<T>(recoilState: RecoilState<T>): [T, (f: (d: Draft<T>) => void) => void] {
  const [st, setState] = useRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper: (t: Immutable<Draft<T>>) => Immutable<Draft<T>> = produce((d) => void f(d));
      // convert Immutable<Draft<T>> to T
      setState((mapper as unknown) as (t: T) => T);
    },
    [setState]
  );
  return [st, setDraft];
}

export function useSetRecoilStateDraft<T>(recoilState: RecoilState<T>): (f: (d: Draft<T>) => void) => void {
  const setState = useSetRecoilState(recoilState);
  const setDraft = React.useCallback(
    function setDraft(f: (d: Draft<T>) => void) {
      const mapper: (t: Immutable<Draft<T>>) => Immutable<Draft<T>> = produce((d) => void f(d));
      // convert Immutable<Draft<T>> to T
      setState((mapper as unknown) as (t: T) => T);
    },
    [setState]
  );
  return setDraft;
}
