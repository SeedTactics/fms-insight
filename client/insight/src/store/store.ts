/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as currentStatus from "../data/current-status";
import * as events from "../data/events";
import * as gui from "../data/gui-state";
import * as routes from "../data/routes";
import * as mat from "../data/material-details";
import * as operators from "../data/operators";
import * as paths from "../data/path-lookup";
import * as serverSettings from "../data/server-settings";
import * as allMatBins from "../data/all-material-bins";
import * as websocket from "./websocket";
import { initBarcodeListener } from "./barcode";

import { createStore, StoreState, StoreActions, ACPayload, mkACF } from "./typed-redux";
import { applyMiddleware, compose } from "redux";
import * as reactRedux from "react-redux";
import { middleware } from "./middleware";

import { connectRoutes, LocationState } from "redux-first-router";
import * as queryString from "query-string";

/* eslint-disable @typescript-eslint/no-use-before-define */

type InitToStore<I> = I extends (a: { useRouter: boolean }) => infer S ? S : never;
export type Store = StoreState<InitToStore<typeof initStore>>;
export type AppActionBeforeMiddleware = StoreActions<InitToStore<typeof initStore>>;
export type Payload<T> = ACPayload<InitToStore<typeof initStore>, T>;
export type DispatchAction<T> = (payload: Payload<T>) => void;
export const connect: InitToStore<typeof initStore>["connect"] = reactRedux.connect;
export const mkAC = mkACF<AppActionBeforeMiddleware>();
export const useSelector: reactRedux.TypedUseSelectorHook<Store> = reactRedux.useSelector;

export function initStore({ useRouter }: { useRouter: boolean }) {
  const router = useRouter
    ? connectRoutes(routes.routeMap, {
        querySerializer: queryString,
      })
    : undefined;

  /* eslint-disable @typescript-eslint/no-explicit-any */
  const composeEnhancers =
    typeof window === "object" && (window as any)["__REDUX_DEVTOOLS_EXTENSION_COMPOSE__"]
      ? (window as any)["__REDUX_DEVTOOLS_EXTENSION_COMPOSE__"]
      : compose;
  /* eslint-enable @typescript-eslint/no-explicit-any */

  const store = createStore(
    {
      Current: currentStatus.reducer,
      Events: events.reducer,
      Gui: gui.reducer,
      MaterialDetails: mat.reducer,
      PathLookup: paths.reducer,
      AllMatBins: allMatBins.reducer,
      Route: routes.reducer,
      Websocket: websocket.reducer,
      Operators: operators.reducer,
      ServerSettings: serverSettings.reducer,
      location: router ? router.reducer : (s: LocationState<string>, _: object) => s || {},
    },
    middleware,
    router
      ? (m) => composeEnhancers(router.enhancer, applyMiddleware(m, router.middleware))
      : (m) => composeEnhancers(applyMiddleware(m))
  );

  initBarcodeListener(store.dispatch.bind(store));

  const operatorOnStateChange = operators.createOnStateChange();
  const matBinsOnStateChange = allMatBins.createOnStateChange();
  store.subscribe(() => {
    operatorOnStateChange(store.getState().Operators);
    matBinsOnStateChange(store.getState().AllMatBins);
  });

  return store;
}
