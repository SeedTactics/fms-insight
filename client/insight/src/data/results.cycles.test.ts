/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { addDays, addHours } from "date-fns";

import { PledgeStatus } from "../store/middleware";
import * as events from "./events";
import { fakeCycle } from "./events.fake";
import { ILogEntry } from "./api";
import {
  stationMinutes,
  filterStationCycles,
  buildCycleTable,
  buildLogEntriesTable,
  outlierLoadCycles,
  outlierMachineCycles,
} from "./results.cycles";

it("creates cycles clipboard table", () => {
  const now = new Date(2018, 2, 5); // midnight in local time

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle(now, 30),
    fakeCycle(addHours(now, -3), 20),
    fakeCycle(addHours(now, -15), 15)
  );
  const st = events.reducer(events.initial, {
    type: events.ActionType.LoadRecentLogEntries,
    now: addDays(now, 1),
    pledge: {
      status: PledgeStatus.Completed,
      result: evts,
    },
  });
  const data = filterStationCycles(st.last30.cycles.part_cycles, undefined, undefined, undefined);

  const table = document.createElement("div");
  table.innerHTML = buildCycleTable(data, true, undefined, undefined);
  expect(table).toMatchSnapshot("cycle clipboard table");

  table.innerHTML = buildCycleTable(data, true, addHours(now, -3), now);
  expect(table).toMatchSnapshot("cycle filtered clipboard table");
});

it("loads outlier cycles", () => {
  const now = new Date(2018, 2, 5); // midnight in local time

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle(now, 30),
    fakeCycle(addHours(now, -3), 20),
    fakeCycle(addHours(now, -15), 15)
  );
  const st = events.reducer(events.initial, {
    type: events.ActionType.LoadRecentLogEntries,
    now: addDays(now, 1),
    pledge: {
      status: PledgeStatus.Completed,
      result: evts,
    },
  });

  const loadOutliers = outlierLoadCycles(
    st.last30.cycles.part_cycles,
    new Date(2018, 0, 1),
    new Date(2018, 11, 1),
    st.last30.cycles.estimatedCycleTimes
  );
  expect(loadOutliers.data.length()).toBe(0);

  const machineOutliers = outlierMachineCycles(
    st.last30.cycles.part_cycles,
    new Date(2018, 0, 1),
    new Date(2018, 11, 1),
    st.last30.cycles.estimatedCycleTimes
  );
  expect(machineOutliers.data.length()).toBe(0);
});

it("computes station oee", () => {
  const now = new Date(2018, 2, 5);

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle(now, 30),
    fakeCycle(addDays(now, -3), 20),
    fakeCycle(addDays(now, -15), 15)
  );
  const st = events.reducer(events.initial, {
    type: events.ActionType.LoadRecentLogEntries,
    now: addDays(now, 1),
    pledge: {
      status: PledgeStatus.Completed,
      result: evts,
    },
  });

  const statMins = stationMinutes(st.last30.cycles.part_cycles, addDays(now, -7));
  expect(statMins).toMatchSnapshot("station minutes for last week");
});

it("creates log entries clipboard table", () => {
  const now = new Date(2018, 2, 5); // midnight in local time

  const evts = ([] as ILogEntry[]).concat(
    fakeCycle(now, 30),
    fakeCycle(addHours(now, -3), 20),
    fakeCycle(addHours(now, -15), 15)
  );
  const table = document.createElement("div");
  table.innerHTML = buildLogEntriesTable(evts);
  expect(table).toMatchSnapshot("events clipboard table");
});
