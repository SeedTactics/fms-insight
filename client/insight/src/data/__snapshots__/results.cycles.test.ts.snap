// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calculates recent cycles > recent cycles 1`] = `
[
  {
    "endActive": 2018-03-05T08:59:00.000Z,
    "endOccupied": 2018-03-05T09:00:00.000Z,
    "outlier": false,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "startTime": 2018-03-05T08:39:00.000Z,
    "station": "L/U #1",
  },
  {
    "endActive": 2018-03-05T09:52:00.000Z,
    "endOccupied": 2018-03-05T10:00:00.000Z,
    "outlier": true,
    "parts": [
      {
        "oper": "LOAD",
        "part": "part3-1",
      },
      {
        "oper": "UNLOAD",
        "part": "part4-2",
      },
    ],
    "startTime": 2018-03-05T09:32:00.000Z,
    "station": "L/U #1",
  },
  {
    "endActive": 2018-03-05T06:59:00.000Z,
    "endOccupied": 2018-03-05T07:00:00.000Z,
    "outlier": false,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1-1",
      },
    ],
    "startTime": 2018-03-05T06:31:00.000Z,
    "station": "MC #1",
  },
  {
    "endActive": 2018-03-05T07:48:00.000Z,
    "endOccupied": 2018-03-05T08:00:00.000Z,
    "outlier": true,
    "parts": [
      {
        "oper": "abcd",
        "part": "part1-1",
      },
    ],
    "startTime": 2018-03-05T07:20:00.000Z,
    "station": "MC #1",
  },
]
`;

exports[`creates cycles clipboard table > cycle clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Part
        </th>
        <th>
          Station
        </th>
        <th>
          Pallet
        </th>
        <th>
          Serial
        </th>
        <th>
          Workorder
        </th>
        <th>
          Inspection
        </th>
        <th>
          Elapsed Min
        </th>
        <th>
          Target Min
        </th>
        <th>
          Median Elapsed Min
        </th>
        <th>
          Median Deviation
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <td>
          Mar 4, 2018, 9:05 AM
        </td>
        <td>
          partz-3
        </td>
        <td>
          L/U #1
        </td>
        <td>
          palw
        </td>
        <td>
          serialH
        </td>
        <td>
          workU
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:28 AM
        </td>
        <td>
          partz-3
        </td>
        <td>
          MC #1
        </td>
        <td>
          palw
        </td>
        <td>
          serialH
        </td>
        <td>
          workU
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          15.0
        </td>
        <td>
          15.0
        </td>
        <td>
          15.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:38 AM
        </td>
        <td>
          partz-3
        </td>
        <td>
          L/U #2
        </td>
        <td>
          palw
        </td>
        <td>
          serialH
        </td>
        <td>
          workU
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:05 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          L/U #1
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:33 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          MC #1
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          20.0
        </td>
        <td>
          20.0
        </td>
        <td>
          20.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:43 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          L/U #2
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:05 AM
        </td>
        <td>
          part7-1
        </td>
        <td>
          L/U #1
        </td>
        <td>
          palH
        </td>
        <td>
          serialP
        </td>
        <td>
          work9
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:43 AM
        </td>
        <td>
          part7-1
        </td>
        <td>
          MC #1
        </td>
        <td>
          palH
        </td>
        <td>
          serialP
        </td>
        <td>
          work9
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          30.0
        </td>
        <td>
          30.0
        </td>
        <td>
          30.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:53 AM
        </td>
        <td>
          part7-1
        </td>
        <td>
          L/U #2
        </td>
        <td>
          palH
        </td>
        <td>
          serialP
        </td>
        <td>
          work9
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;

exports[`creates cycles clipboard table > cycle filtered clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Part
        </th>
        <th>
          Station
        </th>
        <th>
          Pallet
        </th>
        <th>
          Serial
        </th>
        <th>
          Workorder
        </th>
        <th>
          Inspection
        </th>
        <th>
          Elapsed Min
        </th>
        <th>
          Target Min
        </th>
        <th>
          Median Elapsed Min
        </th>
        <th>
          Median Deviation
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <td>
          Mar 4, 2018, 9:05 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          L/U #1
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:33 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          MC #1
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          20.0
        </td>
        <td>
          20.0
        </td>
        <td>
          20.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:43 PM
        </td>
        <td>
          partC-2
        </td>
        <td>
          L/U #2
        </td>
        <td>
          pall
        </td>
        <td>
          serialr
        </td>
        <td>
          workR
        </td>
        <td>
          Insp1[success]
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
        <td>
          0.3
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;

exports[`creates log entries clipboard table > events clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Part
        </th>
        <th>
          Station
        </th>
        <th>
          Pallet
        </th>
        <th>
          Serial
        </th>
        <th>
          Workorder
        </th>
        <th>
          Result
        </th>
        <th>
          Elapsed Min
        </th>
        <th>
          Active Min
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <td>
          Mar 5, 2018, 12:05 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          L/U #1
        </td>
        <td>
          palM
        </td>
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          LOAD
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:10 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          Stocker
        </td>
        <td>
          palM
        </td>
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td />
        <td>
          4.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:43 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          MC #1
        </td>
        <td>
          palM
        </td>
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          progS
        </td>
        <td>
          30.0
        </td>
        <td>
          30.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:53 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          L/U #2
        </td>
        <td>
          palM
        </td>
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          UNLOAD
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:58 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          Signal Insp1
        </td>
        <td />
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          True
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 12:58 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          Signal Insp1
        </td>
        <td />
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          True
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 5, 2018, 1:05 AM
        </td>
        <td>
          part9-3
        </td>
        <td>
          Inspect Insp1
        </td>
        <td />
        <td>
          serial4
        </td>
        <td>
          workb
        </td>
        <td>
          Succeeded
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:05 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          L/U #1
        </td>
        <td>
          palF
        </td>
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          LOAD
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:10 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          Stocker
        </td>
        <td>
          palF
        </td>
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td />
        <td>
          4.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:33 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          MC #1
        </td>
        <td>
          palF
        </td>
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          progH
        </td>
        <td>
          20.0
        </td>
        <td>
          20.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:43 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          L/U #2
        </td>
        <td>
          palF
        </td>
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          UNLOAD
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:48 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          Signal Insp1
        </td>
        <td />
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          True
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:48 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          Signal Insp1
        </td>
        <td />
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          True
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:55 PM
        </td>
        <td>
          partU-2
        </td>
        <td>
          Inspect Insp1
        </td>
        <td />
        <td>
          serialV
        </td>
        <td>
          workj
        </td>
        <td>
          Succeeded
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:05 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          L/U #1
        </td>
        <td>
          pal0
        </td>
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td>
          LOAD
        </td>
        <td>
          6.0
        </td>
        <td>
          6.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:10 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          Stocker
        </td>
        <td>
          pal0
        </td>
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td />
        <td>
          4.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:28 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          MC #1
        </td>
        <td>
          pal0
        </td>
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td>
          progy
        </td>
        <td>
          15.0
        </td>
        <td>
          15.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:38 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          L/U #2
        </td>
        <td>
          pal0
        </td>
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td>
          UNLOAD
        </td>
        <td>
          3.0
        </td>
        <td>
          3.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:43 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          Signal Insp1
        </td>
        <td />
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td>
          True
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

      <tr>
        <td>
          Mar 4, 2018, 9:50 AM
        </td>
        <td>
          part8-1
        </td>
        <td>
          Inspect Insp1
        </td>
        <td />
        <td>
          serialF
        </td>
        <td>
          work4
        </td>
        <td>
          Succeeded
        </td>
        <td>
          0.0
        </td>
        <td>
          0.0
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;
